Sources/NIOCore/VsockAddress.swift

```diff
@@ -0,0 +1,221 @@
+    }
+
+    public init(cid: UInt32, port: UInt32) {
+        var addr = sockaddr_vm()
+        addr.svm_family = sa_family_t(NIOBSDSocket.AddressFamily.vsock.rawValue)
+        addr.svm_cid = cid
+        addr.svm_port = port
+        self.init(address: addr)
+    }
+
+    /// Creates a new vsock address.
+    ///
+    /// - parameters:
+    ///   - cid: the context ID.
+    ///   - port: the target port.
+    /// - returns: the `SocketAddress` for the given context ID and port combination.
+    public init(cid: Int, port: Int) {
+        self.init(cid: UInt32(bitPattern: Int32(truncatingIfNeeded: cid)), port: UInt32(bitPattern: Int32(truncatingIfNeeded: port)))
+    }
+
+    /// Get the local context ID for a vsock socket.
+    ///
+    /// - parameters:
+    ///   - socket: The vsock socket to get the local context ID for.
+    ///
+    /// - NOTE: On Linux, you can use `VMADDR_CID_LOCAL` to retrieve the local context ID, which does not require a socket.
+    public static func localVsockContextID(_ socket: NIOBSDSocket.Handle) -> Int {
+        Int(get_local_vsock_cid(socket))
+    }
+}
+
+extension VsockAddress {
+    /// A vsock Context Identifier (CID).
+    ///
+    /// The CID identifies the source or destination, which is either a virtual machine or the host.
+    public struct ContextID: RawRepresentable, ExpressibleByIntegerLiteral, Hashable, Sendable {
+        public var rawValue: UInt32
+
+        public init(rawValue: UInt32) {
+            self.rawValue = rawValue
+        }
+
+        public init(integerLiteral value: UInt32) {
+            self.init(rawValue: value)
+        }
+
+        /// Wildcard, matches any address.
+        ///
+        /// On all platforms, using this value with `bind(2)` means "any address".
+        ///
+        /// On Darwin platforms, the man page states this can be used with `connect(2)` to mean "this host".
+        public static let any: Self = Self(rawValue: VMADDR_CID_ANY)
+
+        /// The address of the hypervisor.
+        public static let hypervisor: Self = Self(rawValue: UInt32(VMADDR_CID_HYPERVISOR))
+
+        /// The address of the host.
+        public static let host: Self = Self(rawValue: UInt32(VMADDR_CID_HOST))
+
+#if os(Linux)
+        /// The address for local communication (loopback).
+        ///
+        /// This directs packets to the same host that generated them.  This is useful for testing
+        /// applications on a single host and for debugging.
+        ///
+        /// The local context ID obtained with ``getLocalContextID(_:)`` can be used for the same
+        /// purpose, but it is preferable to use ``local``.
+        public static let local: Self = Self(rawValue: UInt32(VMADDR_CID_LOCAL))
+#endif
+
+        /// Get the context ID of the local machine.
+        ///
+        /// On Linux, consider using ``local`` when binding instead of this function.
+        public static func getLocalContextID(_ socket: NIOBSDSocket.Handle) -> Self {
```

Comment:
I don't think we need this function to be public.

Reply:
Is there any reason to allow this public API to be anything other than a channel option?

Reply:
What's the semantics we're trying to achieve that would be covered by a channel option?

IIUC we need this to be something a user can _get_ to support being able to connect to the local CID on platforms that don't have `VMADDR_CID_LOCAL` (Linux < 5.6, Darwin).

If this were a channel option, what address would users provide to `connect(to:)`?

I agree that the primary use case right now is testing, so I'd ideally make it `internal`. What blocks that is the split of the logic over `NIOCore` and `NIOPosix`.

Reply:
We support get-only Channel Options, which are what we're looking for here I think.

Reply:
Added an get-only option in the latest commit. 

Note that the `ChannelOption` is in `NIOPosix` but relies on a function in NIOCore so I had to keep one of these APIs public.

Reply:
I think moving the whole API surface into NIOPosix is the right thing to do.

Reply:
OK, I've now moved everything to `NIOPosix` and made both `getLocalContextID(socketFD:)` and `getLocalContextID()` internal, leaving just `ChannelOptions.localVsockContextID` as the public API for getting the local context ID.